src/cargo/util/progress.rs

```diff
@@ -74,6 +74,112 @@ struct Format {
     style: ProgressStyle,
     max_width: usize,
     max_print: usize,
+    taskbar: TaskbarProgress,
+}
+
+/// Taskbar progressbar
+///
+/// Outputs ANSI sequences according to the `Operating system commands`.
+struct TaskbarProgress {
+    enabled: bool,
+    error: bool,
+}
+
+/// A taskbar progress value printable as ANSI OSC 9;4 escape code
+#[cfg_attr(test, derive(PartialEq, Debug))]
+enum TaskbarValue {
+    /// No output.
+    None,
+    /// Remove progress.
+    Remove,
+    /// Progress value (0-100).
+    Value(f64),
+    /// Indeterminate state (no bar, just animation)
+    Indeterminate,
+    /// Progress value in an error state (0-100).
+    Error(f64),
+}
+
+enum ProgressOutput {
+    /// Print progress without a message
+    PrintNow,
+    /// Progress, message and taskbar progress
+    TextAndTaskbar(String, TaskbarValue),
+    /// Only taskbar progress, no message and no text progress
+    Taskbar(TaskbarValue),
+}
+
+impl TaskbarProgress {
+    #[cfg(test)]
+    fn new(enabled: bool) -> Self {
+        Self {
+            enabled,
+            error: false,
+        }
+    }
+
+    /// Create a `TaskbarProgress` from Cargo's configuration.
+    /// Autodetects support if not explicitly enabled or disabled.
+    fn from_config(gctx: &GlobalContext) -> Self {
+        let enabled = gctx.progress_config().taskbar.unwrap_or_else(|| {
+            // Windows Terminal session.
+            gctx.get_env("WT_SESSION").is_ok()
+            // Compatibility with ConEmu's ANSI support.
+                || gctx.get_env("ConEmuANSI").ok() == Some("ON".into())
+        });
```

Comment:
Can we break the detection out into a function like we do for hyperlinks?

(and this would be allowed to use `std::env::var_os`, like hyperlinks)

Reply:
I'll see what I can do.

> Ideally I see us breaking this out into a crate as some point

After Lennart's push I think that it may be beneficial for the community to have a crate exactly for this one purpose: detecting supported terminal and serializing progress value and status into escape sequences.

Reply:
I've added `cargo::core::shell::supports_progress_report() -> bool`, `cargo::core::shell::Stream::progress_report` and `cargo::core::shell::Shell::progress_report() -> bool`. Hope this wording is clear:

```rust
let enabled = gctx
    .progress_config()
    .taskbar
    .unwrap_or_else(|| gctx.shell().progress_report());
```